flink-tests/src/test/java/org/apache/flink/api/connector/source/lib/DataGeneratorSourceITCase.java

```diff
@@ -0,0 +1,193 @@
+
+        GeneratorFunction<Long, Long> generatorFunction =
+                new GeneratorFunction<Long, Long>() {
+
+                    @Override
+                    public Long map(Long value) {
+                        return value;
+                    }
+                };
+
+        final DataStream<Long> stream = getGeneratorSourceStream(generatorFunction, env, 1_000L);
+
+        final List<Long> result = stream.executeAndCollect(10000);
+
+        assertThat(result).containsExactlyInAnyOrderElementsOf(range(0, 999));
+    }
+
+    @Test
+    @DisplayName("Exceptions from the generator function are not 'swallowed'.")
+    public void testFailingGeneratorFunction() throws Exception {
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(PARALLELISM);
+
+        GeneratorFunction<Long, Long> generatorFunction =
+                value -> {
+                    throw new Exception("boom");
+                };
+
+        final DataStream<Long> stream = getGeneratorSourceStream(generatorFunction, env, 1_000L);
+
+        assertThatThrownBy(
+                        () -> {
+                            stream.executeAndCollect(10000);
+                        })
+                .satisfies(anyCauseMatches("exception on this input:"))
+                .satisfies(anyCauseMatches("boom"));
+    }
+
+    @Test
+    @DisplayName("Exceptions from the generator function initialization are not 'swallowed'.")
+    public void testFailingGeneratorFunctionInitialization() throws Exception {
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(PARALLELISM);
+
+        GeneratorFunction<Long, Long> generatorFunctionFailingInit =
+                new GeneratorFunction<Long, Long>() {
+                    @Override
+                    public void open(SourceReaderContext readerContext) throws Exception {
+                        throw new Exception("boom");
+                    }
+
+                    @Override
+                    public Long map(Long value) {
+                        return value;
+                    }
+                };
+
+        final DataStream<Long> stream =
+                getGeneratorSourceStream(generatorFunctionFailingInit, env, 1_000L);
+
+        // FIX_ME: failure details are swallowed by Flink
+        // Full details are still available at this line:
+        // https://github.com/apache/flink/blob/bccecc23067eb7f18e20bade814be73393401be5/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L758
+        // But the execution falls through to the line below and discards the root cause of
+        // cancelling the source invokable without recording it:
+        // https://github.com/apache/flink/blob/bccecc23067eb7f18e20bade814be73393401be5/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L780
```

Comment:
Please file a ticket; this shouldn't be the case.

Reply:
Done [FLINK-29364](https://issues.apache.org/jira/browse/FLINK-29364 ).